File: vavr/src/test/java/com/github/ljtfreitas/julian/vavr/OptionResponseTTest.java
package com.github.ljtfreitas.julian.vavr;

import com.github.ljtfreitas.julian.Arguments;
import com.github.ljtfreitas.julian.Endpoint;
import com.github.ljtfreitas.julian.JavaType;
import com.github.ljtfreitas.julian.Promise;
import com.github.ljtfreitas.julian.Response;
import com.github.ljtfreitas.julian.ResponseFn;
import io.vavr.control.Option;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OptionResponseTTest {

    @Mock
    private Endpoint endpoint;

    private final OptionResponseT responseT = new OptionResponseT();

    @Nested
    class Predicates {

        @Test
        void supported() {
            when(endpoint.returnType()).thenReturn(JavaType.parameterized(Option.class, String.class));
            assertTrue(responseT.test(endpoint));
        }

        @Test
        void unsupported() {
            when(endpoint.returnType()).thenReturn(JavaType.valueOf(String.class));
            assertFalse(responseT.test(endpoint));
        }
    }

    @Nested
    class Adapt {

        @Test
        void adaptToArgument() {
            when(endpoint.returnType()).thenReturn(JavaType.parameterized(Option.class, String.class));

            JavaType adapted = responseT.adapted(endpoint);

            assertThat(adapted, equalTo(JavaType.valueOf(String.class)));
        }

        @Test
        @DisplayName("adapt to Object when Option is not parameterized")
        void adaptToObjectWhenOptionIsNotParameterized() {
            when(endpoint.returnType()).thenReturn(JavaType.valueOf(Option.class));

            JavaType adapted = responseT.adapted(endpoint);

            assertThat(adapted, equalTo(JavaType.valueOf(Object.class)));
        }
    }

    @Test
    void bind(@Mock Promise<Response<String>> promise, @Mock ResponseFn<String, Object> fn) {
        Arguments arguments = Arguments.empty();

        String content = "hello";

        when(fn.run(promise, arguments)).thenReturn(Promise.done(content));

        Option<Object> option = responseT.bind(endpoint, fn).join(promise, arguments);

        assertTrue(option.exists(content::equals));
    }

    @Test
    void bindNullValue(@Mock Promise<Response<String>> promise, @Mock ResponseFn<String, Object> fn) {
        Arguments arguments = Arguments.empty();

        when(fn.run(promise, arguments)).thenReturn(Promise.done(null));

        Option<Object> option = responseT.bind(endpoint, fn).join(promise, arguments);

        assertTrue(option.isEmpty());
    }
}

File: arrow/src/test/kotlin/com/github/ljtfreitas/julian/k/arrow/OptionResponseTTest.kt
package com.github.ljtfreitas.julian.k.arrow

import arrow.core.Option
import com.github.ljtfreitas.julian.Arguments
import com.github.ljtfreitas.julian.Endpoint
import com.github.ljtfreitas.julian.ObjectResponseT
import com.github.ljtfreitas.julian.Promise
import com.github.ljtfreitas.julian.Response
import com.github.ljtfreitas.julian.k.javaType
import io.kotest.assertions.arrow.core.shouldBeNone
import io.kotest.assertions.arrow.core.shouldBeSome
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk

class OptionResponseTTest : DescribeSpec({

    val subject = OptionResponseT

    val endpoint = mockk<Endpoint>()

    describe("a ResponseT instance for Option<T> values") {

        describe("predicates") {

            it("supports Option<T> as function return type") {
                every { endpoint.returnType() } returns javaType<Option<String>>()

                subject.test(endpoint) shouldBe true
            }

            it("it doesn't support any other return type") {
                every { endpoint.returnType() } returns javaType<String>()

                subject.test(endpoint) shouldBe false
            }
        }

        describe("adapt to expected type") {

            it("we must to adapt to Option argument (Option<T> -> T)") {
                every { endpoint.returnType() } returns javaType<Option<String>>()

                subject.adapted(endpoint) shouldBe javaType<String>()
            }
        }

        describe("bind") {

            every { endpoint.returnType() } returns javaType<Option<String>>()

            val fn = subject.bind<String>(endpoint = endpoint, next = ObjectResponseT<Any>().bind(endpoint, null))

            it("bind a value T to Option<T>") {
                val expected = "hello"

                val option = fn.join(Promise.done(Response.done(expected)), Arguments.empty())

                option shouldBeSome expected
            }

            it("bind a nullable value of type T for a empty Option (a None)") {
                val option = fn.join(Promise.done(Response.empty()), Arguments.empty())

                option.shouldBeNone()
            }
        }
    }
})
